### ISSUE TITLE
How to avoid "Invalid Shares" 

### ISSUE THREAD

**Issue by debuggenius** [fireice-uk/xmr-stak#1776]
Hi there,

I'm making a monero miner bot. First of all thanks for xmr-stak for giving me so many helps. 

Till now I have finished almost all works about mining. But sometime when I submit a hash to the pool, it returns error message "Difficulty Too Low".  However everything is ok when using xmr-stak and I didn't find any code about that in this project. 

What should I do to avoid "Invalid Shares" please? Any help will be appreciated.

**Comment by Spudz76** [comment]
Usually means some sort of algo variant mismatch.  What pool/coin and what algo did you set?

**Comment by debuggenius** [comment]
@Spudz76 monero in supportxmr.

I found that there are only two request from xmr-stak:

login:
SEND: {"method":"login","params":{"login":"xxx","pass":"xxx","rigid":"","agent":"xmr-stak/2.4.7/9c49fb0/master/mac/cpu/20"},"id":1}

submit:
SEND: {"method":"submit","params":{"id":"xxx","job_id":"xxx","nonce":"xxx","result":"xxx"},"id":1}

I have two questions:
1. I can not see any job refresh request and how does xmr-stak get a new job
2. all parameters in xmr-stak's submit request are all same to mine, what's the secret of xmr-stak to make sure that every hash is valid

**Comment by Spudz76** [comment]
You are only seeing the outgoing side, incoming packets will contain a job, which is a nonce and a target difficulty.  You should get one in reply to the login packet, and then more just whenever the pool wants (as jobs rotate / expire).  The "secret" is to check this assignment target to the difficulty of the found result, and if the result is above the target (and the crypto is valid) it's valid and it gets sent.  The CPU double checks the crypto and would say "invalid result" and not even send anything to the pool if that were the failure.  Having sent a result to the pool then the target must have been hit, and the pool is lying or has decided the target should be higher.  Perhaps it is sending a "job change" which is nonstandard and probably not supported - pretty sure we just ignore duplicate job ID# and don't assume it's an update/change but a lie/confusion.

**Comment by psychocrypt** [comment]
Invalid results are in the mist cases the result of overklocked gpus. If
you use the rocm stack combined with amd gpus it could be that there is a
driver issue.
I still looking into the rocm stack why it gives invalid results.
do you overclocked or undervolted your gpu? Or do you use a non standard
bios for the gpu?

**Comment by debuggenius** [comment]
@Spudz76  I scheduled a timer to query new job, and now I keep listening for new job message from the pool, seems that invalid shares reduced. I will spend some time to read your words again and again.  Thank you very much.

**Comment by Spudz76** [comment]
Yes, the "stratum protocol" is not much of a protocol, you will find everyone using it slightly different.
But it is more of a subscription so your login is all the pool needs to start sending you new jobs whenever.
BUT if you send login too much it invalidates previous assignments (basically pool assumes you got disconnected, because why else would you login again?)

**Comment by Spudz76** [comment]
You will learn fastest about the protocol using Wireshark to watch a real mining session.

The "docs" are generally misleading because outdated and most don't have the extensions that have been randomly added/adopted since then.
